File: BusinessSDK/Sources/AdviceSlipSDK/AdviceSlipSDK.swift
import Foundation
import BusinessSDKCore

public enum AdviceSlipSDKConfig {
    public static let baseURL = URL(string: "https://api.adviceslip.com")!
}

public struct AdviceEnvelope: Codable, Sendable {
    public let slip: AdviceSlip

    public init(slip: AdviceSlip) {
        self.slip = slip
    }
}

public struct AdviceSlip: Codable, Sendable {
    public let id: Int
    public let advice: String

    public init(id: Int, advice: String) {
        self.id = id
        self.advice = advice
    }
}


private struct AdviceRepositoryFetchRandomAdviceTESTRequest: APIRequest {


    init() {

    }

    let method: HTTPMethod = .get
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/advice"
    }

    var queryItems: [URLQueryItem] {
        []
    }

    typealias Response = AdviceEnvelope
}

private struct AdviceRepositoryFetchAdviceByIdRequest: APIRequest {
    private let id: Int

    init(id: Int) {
        self.id = id
    }

    let method: HTTPMethod = .get
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/advice/\(id)"
    }

    var queryItems: [URLQueryItem] {
        []
    }

    typealias Response = AdviceEnvelope
}


public protocol AdviceRepositoryProtocol: Sendable {
    func fetchRandomAdviceTEST() async throws -> AdviceEnvelope
    func fetchAdviceById(id: Int) async throws -> AdviceEnvelope
}

public final class AdviceRepository: AdviceRepositoryProtocol, @unchecked Sendable {
    private let networkEngine: NetworkEngineProtocol

    public init(networkEngine: NetworkEngineProtocol) {
        self.networkEngine = networkEngine
    }

public func fetchRandomAdviceTEST() async throws -> AdviceEnvelope {
    try await networkEngine.request(AdviceRepositoryFetchRandomAdviceTESTRequest())
}

public func fetchAdviceById(id: Int) async throws -> AdviceEnvelope {
    try await networkEngine.request(AdviceRepositoryFetchAdviceByIdRequest(id: id))
}

}


public protocol AdviceSlipSDKClientProtocol: Sendable {
    var adviceRepository: AdviceRepositoryProtocol { get }
}

public struct AdviceSlipSDKClient: AdviceSlipSDKClientProtocol, Sendable {
    public let adviceRepository: AdviceRepositoryProtocol

    public init(networkEngine: NetworkEngineProtocol = NetworkEngine(baseURL: AdviceSlipSDKConfig.baseURL)) {
        self.adviceRepository = AdviceRepository(networkEngine: networkEngine)
    }
}

File: android/modules/advice-slip-sdk/src/main/java/com/multiplatformbusinesssdk/adviceslip/AdviceSlipSDK.kt
package com.multiplatformbusinesssdk.adviceslip

import com.multiplatformbusinesssdk.core.ApiRequest
import com.multiplatformbusinesssdk.core.HttpMethod
import com.multiplatformbusinesssdk.core.NetworkEngine

object AdviceSlipSDKConfig {
    const val baseUrl: String = "https://api.adviceslip.com"
}

data class AdviceEnvelope(
    val slip: AdviceSlip
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): AdviceEnvelope {
                        val slip = AdviceSlip.fromJson(json.getJSONObject("slip"))
            return AdviceEnvelope(slip)
        }
    }
}

data class AdviceSlip(
    val id: Int,
    val advice: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): AdviceSlip {
                        val id = json.getInt("id")
            val advice = json.getString("advice")
            return AdviceSlip(id, advice)
        }
    }
}


private class AdviceRepositoryFetchRandomAdviceRequest() : ApiRequest<AdviceEnvelope> {
    override val path: String = "/advice"
    override val method: HttpMethod = HttpMethod.GET
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json")
    override val query: Map<String, String> = emptyMap()
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): AdviceEnvelope =
        AdviceEnvelope.fromJson(org.json.JSONObject(payload) as org.json.JSONObject)
}

private class AdviceRepositoryFetchAdviceByIdRequest(private val id: Int) : ApiRequest<AdviceEnvelope> {
    override val path: String = "/advice/$id"
    override val method: HttpMethod = HttpMethod.GET
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json")
    override val query: Map<String, String> = emptyMap()
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): AdviceEnvelope =
        AdviceEnvelope.fromJson(org.json.JSONObject(payload) as org.json.JSONObject)
}


interface AdviceRepositoryProtocol {
    suspend fun fetchRandomAdvice(): AdviceEnvelope
    suspend fun fetchAdviceById(id: Int): AdviceEnvelope
}

class AdviceRepository(
    private val networkEngine: NetworkEngine
) : AdviceRepositoryProtocol {
override suspend fun fetchRandomAdvice(): AdviceEnvelope {
    return networkEngine.request(AdviceRepositoryFetchRandomAdviceRequest())
}

override suspend fun fetchAdviceById(id: Int): AdviceEnvelope {
    return networkEngine.request(AdviceRepositoryFetchAdviceByIdRequest(id = id))
}

}


interface AdviceSlipSDKClientProtocol {
    val adviceRepository: AdviceRepositoryProtocol
}

class AdviceSlipSDKClient(
    networkEngine: NetworkEngine = NetworkEngine(baseUrl = AdviceSlipSDKConfig.baseUrl)
) : AdviceSlipSDKClientProtocol {
    override val adviceRepository: AdviceRepositoryProtocol = AdviceRepository(networkEngine)
}
